lib/scholar/cluster/gmm.ex

```diff
@@ -0,0 +1,407 @@
+          if max_lower_bound < lower_bound do
+            {lower_bound, params}
+          else
+            {max_lower_bound, best_params}
+          end
+
+        {
+          best_params,
+          {run + 1, max_lower_bound, x, key, max_iter, tol, reg_covar}
+        }
+      end
+
+    {weights, means, covariances, precisions_cholesky} = best_params
+
+    %__MODULE__{
+      weights: weights,
+      means: means,
+      covariances: covariances,
+      precisions_cholesky: precisions_cholesky
+    }
+  end
+
+  defnp init_params(key, x, opts) do
+    type = Nx.type(x)
+    {num_samples, _} = Nx.shape(x)
+    num_gaussians = opts[:num_gaussians]
+    reg_covar = opts[:reg_covar]
+
+    k_means_model =
+      Scholar.Cluster.KMeans.fit(x, num_clusters: num_gaussians, num_runs: 1, key: key)
+
+    labels = k_means_model.labels
+    resp = Nx.take(Nx.eye(num_gaussians, type: type), labels)
+
+    {weights, means, covariances, precisions_cholesky} =
+      estimate_gaussian_parameters(x, resp, reg_covar)
+
+    weights = weights / num_samples
+    {weights, means, covariances, precisions_cholesky}
+  end
+
+  defnp e_step(x, weights, means, precisions_cholesky) do
+    {log_prob_norm, log_resp} = estimate_log_prob_resp(x, weights, means, precisions_cholesky)
+    {Nx.mean(log_prob_norm), log_resp}
+  end
+
+  defnp m_step(x, log_resp, reg_covar) do
+    {weights, means, covariances, precisions_cholesky} =
+      estimate_gaussian_parameters(x, Nx.exp(log_resp), reg_covar)
+
+    weights = weights / Nx.sum(weights)
+    {weights, means, covariances, precisions_cholesky}
+  end
+
+  defnp estimate_gaussian_parameters(x, resp, reg_covar) do
+    nk = Nx.sum(resp, axes: [0])
+    means = Nx.dot(Nx.transpose(resp), x) / nk
+    covariances = estimate_gaussian_covariances(x, resp, nk, means, reg_covar)
+    precisions_cholesky = compute_precisions_cholesky(covariances)
+    {nk, means, covariances, precisions_cholesky}
+  end
+
+  defnp estimate_gaussian_covariances(x, resp, nk, means, reg_covar) do
+    type = Nx.type(x)
+    {num_gaussians, num_features} = Nx.shape(means)
+
+    {covariances, _} =
+      while {
+              covariances =
+                Nx.tensor(0.0, type: type)
+                |> Nx.broadcast({num_gaussians, num_features, num_features}),
+              {k = 0, x, resp, nk, means, reg_covar}
+            },
+            k < num_gaussians do
+        diff = x - means[k]
+        covariance = Nx.dot(resp[[.., k]] * Nx.transpose(diff), diff) / nk[k]
+        covariance = Nx.put_diagonal(covariance, Nx.take_diagonal(covariance) + reg_covar)
+
+        {
+          Nx.put_slice(covariances, [k, 0, 0], Nx.new_axis(covariance, 0)),
+          {k + 1, x, resp, nk, means, reg_covar}
+        }
+      end
+
+    covariances
+  end
+
+  defnp compute_precisions_cholesky(covariances) do
+    type = Nx.type(covariances)
+    {num_gaussians, num_features, _} = Nx.shape(covariances)
+
+    {precisions_cholesky, _} =
+      while {
```

Comment:
Same thing about vectorization here

Reply:
Yeah, this one might be easier as both `Nx.LinAlg.cholesky` and `Nx.LinAlg.triangular_solve` work with batches of matrices.

Reply:
You don't have to think much about this. Nx main has Nx.vectorize/devectorize/revectorize which will handle batching for you

Reply:
One thing was strange to me. Variable `type` is not declared in the `while` loop and it still compiles. Is this intended?

Reply:
In fact, I'm not sure it _would_ compile if you passed type as an argument :)

type is a tuple of {atom, number} used only at compile time. Passing as argument would force its traversal as an Nx.Container and that would break on the Atom.

It's ok anyway because the type is just used at compile-time

Reply:
Nx only checks the arguments between def <-> defn boundaries.